File: services/client/src/protocol/bet_record.go
package protocol

import (
	"encoding/binary"
	"fmt"
	"math"
	"unicode/utf8"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/business"
)

const (
	documentBytes   = 4
	numberBytes     = 2
	birthdateBytes  = 10
	nameLengthBytes = 1

	fixedPrefixBytes   = documentBytes + numberBytes + birthdateBytes
	minimumRecordBytes = fixedPrefixBytes + nameLengthBytes
)

func betToBytes(bet business.Bet) ([]byte, error) {
	if int64(bet.Document) > math.MaxUint32 {
		return nil, fmt.Errorf("document %d does not fit in %d bytes", bet.Document, documentBytes)
	}
	if bet.Number > math.MaxUint16 {
		return nil, fmt.Errorf("number %d does not fit in %d bytes", bet.Number, numberBytes)
	}
	if len(bet.Birthdate) != birthdateBytes {
		return nil, fmt.Errorf(
			"birthdate must be exactly %d bytes, got %d: %q",
			birthdateBytes, len(bet.Birthdate), bet.Birthdate,
		)
	}
	if len(bet.FirstName) > math.MaxUint8 {
		return nil, fmt.Errorf(
			"first_name is %d bytes, at most %d fit in the length prefix",
			len(bet.FirstName), math.MaxUint8,
		)
	}
	if len(bet.LastName) > math.MaxUint8 {
		return nil, fmt.Errorf(
			"last_name is %d bytes, at most %d fit in the length prefix",
			len(bet.LastName), math.MaxUint8,
		)
	}

	totalRecordBytes := fixedPrefixBytes + 2*nameLengthBytes + len(bet.FirstName) + len(bet.LastName)

	recordBytes := make([]byte, 0, totalRecordBytes)
	recordBytes = binary.BigEndian.AppendUint32(recordBytes, uint32(bet.Document))
	recordBytes = binary.BigEndian.AppendUint16(recordBytes, uint16(bet.Number))
	recordBytes = append(recordBytes, bet.Birthdate...)
	recordBytes = append(recordBytes, byte(len(bet.FirstName)))
	recordBytes = append(recordBytes, bet.FirstName...)
	recordBytes = append(recordBytes, byte(len(bet.LastName)))
	recordBytes = append(recordBytes, bet.LastName...)

	return recordBytes, nil
}

func betFromBytes(payload []byte) (business.Bet, error) {
	if len(payload) < minimumRecordBytes {
		return business.Bet{}, fmt.Errorf(
			"bet record needs at least %d bytes, got %d",
			minimumRecordBytes, len(payload),
		)
	}

	offset := 0
	document := binary.BigEndian.Uint32(payload[offset : offset+documentBytes])
	offset += documentBytes

	number := binary.BigEndian.Uint16(payload[offset : offset+numberBytes])
	offset += numberBytes

	birthdate, err := decodedText(payload[offset:offset+birthdateBytes], "birthdate")
	if err != nil {
		return business.Bet{}, err
	}
	offset += birthdateBytes

	firstNameLength := int(payload[offset])
	offset += nameLengthBytes
	if offset+firstNameLength+nameLengthBytes > len(payload) {
		return business.Bet{}, fmt.Errorf(
			"first_name length %d overruns the record: only %d bytes remain",
			firstNameLength, len(payload)-offset,
		)
	}
	firstName, err := decodedText(payload[offset:offset+firstNameLength], "first_name")
	if err != nil {
		return business.Bet{}, err
	}
	offset += firstNameLength

	lastNameLength := int(payload[offset])
	offset += nameLengthBytes
	if offset+lastNameLength != len(payload) {
		return business.Bet{}, fmt.Errorf(
			"last_name length %d does not consume the record exactly: %d bytes remain",
			lastNameLength, len(payload)-offset,
		)
	}
	lastName, err := decodedText(payload[offset:offset+lastNameLength], "last_name")
	if err != nil {
		return business.Bet{}, err
	}

	return business.Bet{
		Document:  int(document),
		Number:    int(number),
		Birthdate: birthdate,
		FirstName: firstName,
		LastName:  lastName,
	}, nil
}

// python bytes.decode("utf-8") raises on invalid bytes, since go's doesn't have this capability
// placing this helper here to handle that scenario
func decodedText(raw []byte, field string) (string, error) {
	if !utf8.Valid(raw) {
		return "", fmt.Errorf("%s is not valid utf-8", field)
	}
	return string(raw), nil
}

File: services/server/src/protocol/messages/bet_record.py
from lottery.bet import Bet
from protocol.errors import ProtocolError


class BetRecord:
    _DOCUMENT_BYTES = 4
    _NUMBER_BYTES = 4
    _BIRTHDATE_BYTES = 10
    _NAME_LENGTH_BYTES = 1
    _FIXED_PREFIX_BYTES = _DOCUMENT_BYTES + _NUMBER_BYTES + _BIRTHDATE_BYTES

    def __init__(self, document, number, birthdate, first_name, last_name):
        birthdate_in_bytes = birthdate.encode("utf-8")
        if len(birthdate_in_bytes) != self._BIRTHDATE_BYTES:
            raise ProtocolError(
                f"birthdate must be exactly {self._BIRTHDATE_BYTES} bytes, "
                f"got {len(birthdate_in_bytes)}: {birthdate!r}"
            )
        self._document = document
        self._number = number
        self._birthdate = birthdate
        self._first_name = first_name
        self._last_name = last_name

    @classmethod
    def from_bet(cls, bet: Bet):
        return cls(
            document=bet.document,
            number=bet.number,
            birthdate=bet.birthdate,
            first_name=bet.first_name,
            last_name=bet.last_name,
        )

    @classmethod
    def from_bytes(cls, data):
        record, consumed = cls.consume_from(data)
        if consumed != len(data):
            raise ProtocolError(
                f"bet record does not consume the payload exactly: "
                f"{len(data) - consumed} bytes remain"
            )
        return record

    @classmethod
    def consume_from(cls, data, offset=0):
        minimum = cls._FIXED_PREFIX_BYTES + 2 * cls._NAME_LENGTH_BYTES
        available = len(data) - offset
        if available < minimum:
            raise ProtocolError(
                f"bet record needs at least {minimum} bytes, got {available}"
            )

        document = int.from_bytes(
            data[offset : offset + cls._DOCUMENT_BYTES], byteorder="big"
        )
        offset += cls._DOCUMENT_BYTES
        number = int.from_bytes(data[offset : offset + cls._NUMBER_BYTES], byteorder="big")
        offset += cls._NUMBER_BYTES
        birthdate = cls._decoded(
            data[offset : offset + cls._BIRTHDATE_BYTES], "birthdate"
        )
        offset += cls._BIRTHDATE_BYTES
        first_name_length = data[offset]
        offset += cls._NAME_LENGTH_BYTES
        if offset + first_name_length + cls._NAME_LENGTH_BYTES > len(data):
            raise ProtocolError(
                f"first_name length {first_name_length} overruns the record: "
                f"only {len(data) - offset} bytes remain"
            )
        first_name = cls._decoded(
            data[offset : offset + first_name_length], "first_name"
        )
        offset += first_name_length
        last_name_length = data[offset]
        offset += cls._NAME_LENGTH_BYTES
        if offset + last_name_length > len(data):
            raise ProtocolError(
                f"last_name length {last_name_length} overruns the record: "
                f"only {len(data) - offset} bytes remain"
            )
        last_name = cls._decoded(
            data[offset : offset + last_name_length], "last_name"
        )
        offset += last_name_length
        return cls(document, number, birthdate, first_name, last_name), offset

    @classmethod
    def _decoded(cls, raw, field):
        try:
            return raw.decode("utf-8")
        except UnicodeDecodeError as e:
            raise ProtocolError(f"{field} is not valid utf-8") from e

    def to_bytes(self):
        first_name_in_bytes = self._first_name.encode("utf-8")
        last_name_in_bytes = self._last_name.encode("utf-8")
        return (
            self._document.to_bytes(self._DOCUMENT_BYTES, byteorder="big")
            + self._number.to_bytes(self._NUMBER_BYTES, byteorder="big")
            + self._birthdate.encode("utf-8")
            + len(first_name_in_bytes).to_bytes(self._NAME_LENGTH_BYTES, byteorder="big")
            + first_name_in_bytes
            + len(last_name_in_bytes).to_bytes(self._NAME_LENGTH_BYTES, byteorder="big")
            + last_name_in_bytes
        )

    def to_bet(self, agency_id: int) -> Bet:
        return Bet(
            agency_id=agency_id,
            first_name=self._first_name,
            last_name=self._last_name,
            document=self._document,
            birthdate=self._birthdate,
            number=self._number,
        )
